### PR TITLE
Planner: Avoid injecting redundant projection below LogicalJoin

### LINKED ISSUE #11650: Planner: Avoid injecting redundant projection when build LogicalJoin
## Feature Request

**Is your feature request related to a problem? Please describe:**

``` mysql
mysql> explain select /*+ TIDB_INLJ(t1) */ t1.a from t t1, t t2 where t1.a = t2.a+1;
+--------------------------+----------+------+----------------------------------------------------------------------------+
| id                       | count    | task | operator info                                                              |
+--------------------------+----------+------+----------------------------------------------------------------------------+
| Projection_8             | 12500.00 | root | test.t1.a                                                                  |
| └─HashLeftJoin_9         | 12500.00 | root | inner join, inner:Projection_14, equal:[eq(test.t1.a, plus(test.t2.a, 1))] |
|   ├─TableReader_13       | 10000.00 | root | data:TableScan_12                                                          |
|   │ └─TableScan_12       | 10000.00 | cop  | table:t1, range:[-inf,+inf], keep order:false, stats:pseudo                |
|   └─Projection_14        | 10000.00 | root | test.t2.a, plus(test.t2.a, 1)                                              |
|     └─TableReader_16     | 10000.00 | root | data:TableScan_15                                                          |
|       └─TableScan_15     | 10000.00 | cop  | table:t2, range:[-inf,+inf], keep order:false, stats:pseudo                |
+--------------------------+----------+------+----------------------------------------------------------------------------+
7 rows in set, 1 warning (0.00 sec)

mysql> show warnings;
+---------+------+-----------------------------------------------------+
| Level   | Code | Message                                             |
+---------+------+-----------------------------------------------------+
| Warning | 1815 | Optimizer Hint /*+ TIDB_INLJ(t1) */ is inapplicable |
+---------+------+-----------------------------------------------------+
1 row in set (0.00 sec)
```

In this case, t1 should  be used as inner table. But the equal condition is `t1.a = t2.a+1`,  we will inject projections at both sides. `Projection(t1.a -> t1.a)` will be eliminated in the last step, but it still exists when considering `IndexJoin` and projection can't be used as inner table.  

**Describe the feature you'd like:**

Avoid injecting meaningless projection when build LogicalJoin.

### PR DESCRIPTION
**Pull request by foreyes**
### What problem does this PR solve? 

Fix #11650 

### What is changed and how it works?

Check whether the projection is necessary, and only inject a necessary projection. 
After this change, `IndexJoin` can be used in this case:
``` mysql
mysql> explain select /*+ TIDB_INLJ(t1) */ t1.a from t t1, t t2 where t1.a = t2.a+1;
+--------------------------+----------+------+------------------------------------------------------------------------------------+
| id                       | count    | task | operator info                                                                      |
+--------------------------+----------+------+------------------------------------------------------------------------------------+
| Projection_7             | 12500.00 | root | test.t1.a                                                                          |
| └─IndexJoin_10           | 12500.00 | root | inner join, inner:TableReader_9, outer key:plus(test.t2.a, 1), inner key:test.t1.a |
|   ├─TableReader_9        | 1.00     | root | data:TableScan_8                                                                   |
|   │ └─TableScan_8        | 1.00     | cop  | table:t1, range: decided by [plus(test.t2.a, 1)], keep order:false, stats:pseudo   |
|   └─Projection_11        | 10000.00 | root | test.t2.a, plus(test.t2.a, 1)                                                      |
|     └─TableReader_13     | 10000.00 | root | data:TableScan_12                                                                  |
|       └─TableScan_12     | 10000.00 | cop  | table:t2, range:[-inf,+inf], keep order:false, stats:pseudo                        |
+--------------------------+----------+------+------------------------------------------------------------------------------------+
7 rows in set (0.00 sec)
```

### Check List 

Tests 

 - Unit test

Code changes

Side effects

 - Change optimizer behaviors.

Related changes

### CODE REVIEW

**Comment by foreyes** [comment]
@winoros Because rule ppd is after rule project eliminating... I have no idea what will happen if we change the rule order.

**Comment by foreyes** [comment]
@lzmhhh123  Looks like it is. I tired `from t t1, tt t2`, still can't choose `IndexJoin` before this PR.

**Comment by lzmhhh123** [comment]
@foreyes That's not my means. Look at the following case:
```sql
tidb(localhost:4000) > select * from mysql.opt_rule_blacklist;
+----------------------+
| name                 |
+----------------------+
| projection_eliminate |
+----------------------+
1 row in set (0.00 sec)

tidb(localhost:4000) > desc select /*+ tidb_inlj(t1, t2) */ * from tt t1 join tt t2 on t1.a = t2.a+1;
+--------------------------+----------+------+----------------------------------------------------------------------------+
| id                       | count    | task | operator info                                                              |
+--------------------------+----------+------+----------------------------------------------------------------------------+
| Projection_7             | 12500.00 | root | test.t1.a, test.t1.b, test.t2.a, test.t2.b                                 |
| └─HashLeftJoin_8         | 12500.00 | root | inner join, inner:Projection_13, equal:[eq(test.t1.a, plus(test.t2.a, 1))] |
|   ├─TableReader_12       | 10000.00 | root | data:TableScan_11                                                          |
|   │ └─TableScan_11       | 10000.00 | cop  | table:t1, range:[-inf,+inf], keep order:false, stats:pseudo                |
|   └─Projection_13        | 10000.00 | root | test.t2.a, test.t2.b, plus(test.t2.a, 1)                                   |
|     └─TableReader_15     | 10000.00 | root | data:TableScan_14                                                          |
|       └─TableScan_14     | 10000.00 | cop  | table:t2, range:[-inf,+inf], keep order:false, stats:pseudo                |
+--------------------------+----------+------+----------------------------------------------------------------------------+
7 rows in set, 1 warning (0.00 sec)

tidb(localhost:4000) > show create table tt;
+-------+---------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                            |
+-------+---------------------------------------------------------------------------------------------------------------------------------------------------------+
| tt    | CREATE TABLE `tt` (
  `a` int(11) NOT NULL,
  `b` int(11) DEFAULT NULL,
  PRIMARY KEY (`a`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+---------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

Why can't it consider `t1` as the inner plan? It looks the projection push down doesn't affect the plan about `t1` side of the join.

**Comment by foreyes** [comment]
@lzmhhh123 In fact, there is a `projection` above `t1`, it's not shown in the plan because it has been eliminated after physical optimize, not by `projection_eliminate`. You can take a look: [https://github.com/pingcap/tidb/blob/master/planner/core/optimizer.go#L134](url)

File: planner/core/rule_predicate_push_down.go [hunk 1]
```diff
@@ -218,12 +218,35 @@ func (p *LogicalJoin) updateEQCond() {
...
+			if !needLProj {
+				lKey, _ = lKeys[i].(*expression.Column)
+			} else {
+				lKey = lProj.appendExpr(lKeys[i])
+			}
+			if !needRProj {
+				rKey, _ = rKeys[i].(*expression.Column)
+			} else {
+				rKey = rProj.appendExpr(rKeys[i])
+			}
```
Comment (eurekaka) [inline planner/core/rule_predicate_push_down.go]: How about:
```golang
lKey, rKey := lKeys[i], rKeys[i]
if lProj != nil {
    lKey = lProj.appendExpr(lKey)
}
if rProj != nil {
    rKey = rProj.appendExpr(rKey)
}
```

**Pull request merged**